File: 卡码/代码随想录/1/10-单调栈/Java/src/com/kurumi/LeetCode_0503.java
//503. 下一个更大元素 II
package com.kurumi;

import java.util.Arrays;
import java.util.Stack;

public class LeetCode_0503 {

}
class Solution {
  public int[] nextGreaterElements(int[] nums) {
  	int[] res=new int[nums.length];
  	Arrays.fill(res,-1);
  	Stack<Integer> st=new Stack<>();
  	for(int i=0;i<nums.length*2;i++) {
  		while(!st.empty()&&nums[i%nums.length]>nums[st.peek()]) {
  			res[st.peek()]=nums[i%nums.length];
  			st.pop();
  		}
  		st.push(i%nums.length);
  	}
  	return res;
  }
}

File: 卡码/代码随想录/1/10-单调栈/Go/LeetCode_0503.go
// 503. 下一个更大元素 II
package main

func nextGreaterElements(nums []int) []int {
	res := make([]int, len(nums))
	for i := 0; i < len(nums); i++ {
		res[i] = -1
	}
	st := make([]int, 0)
	for i := 0; i < len(nums)*2; i++ {
		for len(st) > 0 && nums[i%len(nums)] > nums[st[len(st)-1]] {
			res[st[len(st)-1]] = nums[i%len(nums)]
			st = st[:len(st)-1]
		}
		st = append(st, i%len(nums))
	}
	return res
}
